cryptol-remote-api/src/CryptolServer/Names.hs

```diff
@@ -47,46 +52,80 @@ instance Doc.DescribedMethod VisibleNamesParams [NameInfo] where
       Doc.Paragraph [ Doc.Text " A"
                     , Doc.Link (Doc.TypeDesc (typeRep (Proxy @JSONSchema))) "JSON Cryptol type"
                     ])
+    , ("module",
+      Doc.Paragraph [Doc.Text "A human-readable representation of the module from which the name originates"])
+    , ("parameter",
+      Doc.Paragraph [ Doc.Text "An optional field which is present and ",
+                      Doc.Literal "True", Doc.Text " iff the name is a module parameter" ])
```

Comment:
The convention of _only_ including the `parameter` field if it is equal to `True` is a bit odd, since this means that users would have to write code like this:

```py
if 'parameter' in x:
```

Rather than:

```py
if x['parameter']:
```

That is, the actual value of the boolean is never useful, just the presence or absence of the field itself. This might come down to personal taste, but I would favor always including the `parameter` field, even if it is `False`.

Similar reasoning applies to the `infix` and `pragmas` fields, which are also not included under certain circumstances.

Reply:
This is an excellent point and something I wasn't sure about. First of all, I got the idea from the existing codebase, since this is how the existing `documentation` parameter works: when there is no documentation, the `documentation` parameter simply isn't there.

But my main reason came from remembering that in the end everything gets sent as a JSON string through the RPC interface. There are 125 Prelude definitions that get sent over `names()`, so if every one of them has "`"parameter: False", `" (20 characters), for example, then that's 2,500 characters of redundant information sent every time `names()` is called. Now, I don't know if we even care about that, 2/3K is really not that much data. But it was enough to sway me in the moment.

I don't have strong feelings, so if you don't think we care about how big of a string we send, then I'd be very happy to switch to the (admittedly, easier to use) interface you suggest.

I suppose a compromise would be to keep the RPC API as-is, but on the python side add `parameter: False` whenever `parameter` is not present, and so on. Again, no strong feelings, so I'm curious what you think.

Reply:
I'm not particularly bothered by the fact that we omit fields in certain situations, as that is a pretty natural way to encode a `Maybe` type in an RPC context. That is, omitting the field represents `Nothing`, and including the field represents `Just`. The `documentation` field, for instance, is a `Maybe String`, and the way it's encoded matches my expectations about how `Maybe`s work in RPC.

The thing that bothers me here is that `parameter` _isn't_ a `Maybe`, but rather a raw `Bool`. Except that it's not encoded like a typical RPC `Bool`: in this case, `False` is encoded by omitting the field, and `True` is encoded by including `parameter: True`. In effect, we've made `parameter: False` an unrepresentable state.

Then again, your point about RPC message length is a very compelling argument, and I don't have a good counterargument. In that case, I relent :)

Reply:
Oh, true – that's an excellent point about `Maybe`! So I guess if I want this behavior, I should really encode `parameter` as a `Maybe ()`, i.e. in JSON it's either present and `[]` or not present. I think I'm going to do that and make it clear in the documentation the right way to check if a name is a parameter is to do `"parameter" in n`. This gets rid of that unused `False` state.

Reply:
What I settled on is `parameter` is always either unit or not present. In the JSON this means `[]` or not present, and in the Python this means `()` and not present.